lib/Conversion/ImportVerilog/Structure.cpp

```diff
@@ -94,15 +94,110 @@ struct MemberVisitor {
   // Skip typedefs.
   LogicalResult visit(const slang::ast::TypeAliasType &) { return success(); }
 
+  // Skip ports which are already handled by the module itself.
+  LogicalResult visit(const slang::ast::PortSymbol &) { return success(); }
+  LogicalResult visit(const slang::ast::MultiPortSymbol &) { return success(); }
+
   // Handle instances.
   LogicalResult visit(const slang::ast::InstanceSymbol &instNode) {
-    auto targetModule = context.convertModuleHeader(&instNode.body);
-    if (!targetModule)
+    using slang::ast::ArgumentDirection;
+    using slang::ast::AssignmentExpression;
+    using slang::ast::MultiPortSymbol;
+    using slang::ast::PortSymbol;
+
+    auto *moduleLowering = context.convertModuleHeader(&instNode.body);
+    if (!moduleLowering)
+      return failure();
+    auto module = moduleLowering->op;
+    auto moduleType = module.getModuleType();
+
+    // Prepare the values that are involved in port connections. This creates
+    // rvalues for input ports and appropriate lvalues for output, inout, and
+    // ref ports. We also separate multi-ports into the individual underlying
+    // ports with their corresponding connection.
```

Comment:
From my perspective, in your case(module PortsAnsi), `input`--`%a`, `inout`--`%c`, and `ref`--`%d` are all regarded as the rhs of `moore.assign`. And only `output`--`%d` is the operand of `moore.output`.

Reply:
Yes absolutely! :slightly_smiling_face: My idea was to make almost everything an input and then use your cool new `ref<T>` type that you've been working on:

- `input T` becomes `T` input
- `inout T` becomes `ref<T>` input (maybe later `net<T>` if we decide we need to distinguish vars and nets)
- `ref T` becomes `ref<T>` input
- `output T` becomes `T` output

At the instance, the input would be lowered as rvalue to get a `T`, and all others would be lowered as lvalue to get a `ref<T>`. The refs for inout and ref ports are passed directly to the instance as operands, and outputs generate a `moore.assign` with the lvalue on the left and the instance's output value on the right:

```mlir
%a = moore.variable 
%0 = moore.read %a  // rvalue for `a`
%b = moore.variable  // lvalue for `b`
%c = moore.net  // lvalue for `c`
%d = moore.variable  // lvalue for `d`
%1 = moore.instance ...(a: %0: T, c: %c: net<T>, d: %d: ref<T>) -> (b: T)
moore.assign %b, %1
```